File: Data Structures & Algorithms/subsets/submission-11.js
class Solution {
    /**
     * @param {number[]} nums
     * @return {number[][]}
     */
    subsets(nums) { 
        const res = []
        function dfs(start, cur) {


                res.push([...cur]) 


            for (let i = start; i < nums.length; i ++) {
                cur.push(nums[i]) 
                dfs(i + 1, cur) 
                cur.pop()
            }      
        } 

        dfs(0, []) 

        return res
    }
}

File: Data Structures & Algorithms/generate-parentheses/submission-11.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        res = []
        def dfs(start, end, cur): 
            if start == end and start == n: 
                res.append(cur) 
                return 
            
            if start < n:
                dfs(start + 1, end, cur + "(") 

            if start > end: 
                dfs(start, end + 1, cur + ")")  

        dfs(0, 0, "") 
        return res
            